File: Python/src/linkedlist/find_middle_node.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class FindMiddleNode:

    def middle_node(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow_pointer = fast_pointer = head
        while fast_pointer.next is not None:
            slow_pointer = slow_pointer.next
            fast_pointer = fast_pointer.next
            if fast_pointer.next is not None:
                fast_pointer = fast_pointer.next
        return slow_pointer

File: Cpp/src/linkedlist/find_middle_node.cpp
#include <iostream>
#include "find_middle_node.h"

using namespace std;

ListNode *FindMiddleNode::MiddleNode(ListNode *head) {
    ListNode *slow_pointer;
    ListNode *fast_pointer;
    if (head->next == nullptr) {
        return head;
    }
    slow_pointer = fast_pointer = head;

    while (fast_pointer->next) {
        slow_pointer = slow_pointer->next;
        fast_pointer = fast_pointer->next;
        if (fast_pointer->next) {
            fast_pointer = fast_pointer->next;
        }
    }

    return slow_pointer;
}
